ambari-server/src/main/java/org/apache/ambari/server/checks/ServicesUpCheck.java

```diff
@@ -83,117 +84,124 @@ public ServicesUpCheck() {
+  
+        Map<String, ServiceComponent> serviceComponents = service.getServiceComponents();
+        for (Map.Entry<String, ServiceComponent> component : serviceComponents.entrySet()) {
+  
+          ServiceComponent serviceComponent = component.getValue();
+          // In Services like HDFS, ignore components like HDFS Client
+          if (serviceComponent.isClientComponent()) {
+            continue;
           }
-        }
-
-        // check threshold for slaves which have a non-0 cardinality
-        if (checkThreshold) {
-          int total = hostComponentSummaries.size();
-          int up = 0;
-          int down = 0;
-
-          for (HostComponentSummary summary : hostComponentSummaries) {
-            if (isConsideredDown(cluster, serviceComponent, summary)) {
-              down++;
-            } else {
-              up++;
-            }
+  
+          // skip if the component is not part of the finalization version check
+          if (!serviceComponent.isVersionAdvertised()) {
+            continue;
           }
-
-          if ((float) down / total > SLAVE_THRESHOLD) { // arbitrary
-            failedServices.add(new ServiceDetail(serviceName));
-
-            String message = MessageFormat.format(
-                "{0}: {1} out of {2} {3} are started; there should be {4,number,percent} started before upgrading.",
-                service.getName(), up, total, serviceComponent.getName(), SLAVE_THRESHOLD);
-            errorMessages.add(message);
+  
+          // TODO, add more logic that checks the Upgrade Pack.
+          // These components are not in the upgrade pack and do not advertise a
+          // version:
+          // ZKFC, Ambari Metrics, Kerberos, Atlas (right now).
+          // Generally, if it advertises a version => in the upgrade pack.
+          // So it can be in the Upgrade Pack but not advertise a version.
+          List<HostComponentSummary> hostComponentSummaries = HostComponentSummary.getHostComponentSummaries(
+            service.getClusterId(), service.getServiceGroupId(), service.getServiceId(), serviceComponent.getName());
+  
+          // not installed, nothing to do
+          if (hostComponentSummaries.isEmpty()) {
+            continue;
+          }
+  
+          // non-master, "true" slaves with cardinality 1+
+          boolean checkThreshold = false;
+          if (!serviceComponent.isMasterComponent()) {
+            StackId stackId = service.getStackId();
+            ComponentInfo componentInfo = ambariMetaInfo.get().getComponent(stackId.getStackName(),
+                stackId.getStackVersion(), serviceComponent.getServiceType(),
+                serviceComponent.getName());
+  
+            String cardinality = componentInfo.getCardinality();
+            // !!! check if there can be more than one. This will match, say,
+            // datanodes but not ZKFC
+            if (null != cardinality
+                && (cardinality.equals("ALL") || cardinality.matches("[1-9].*"))) {
```

Comment:
These magic string/magic regex comparisonsons look unreliable in a long term perspective. How about scheduling refactoring to incapsulate Cardinality in a dedicated class with pre-defined constants like ALL and utility methods like moreThenZero()

Reply:
I didn't change these - when we revisit upgrade checks, we can address workarounds like this then.